Feature/add solidity uint types

correcthorsebatterystaple opened the pull request:
Add uint variants for solidity and ensure that all the types behave with literals when arithmetic operations are performed

correcthorsebatterystaple committed:
Add all uint types

correcthorsebatterystaple committed:
Fix literal type missmatch

correcthorsebatterystaple committed:
Default literal type to the type description of literal when passed in type is missing

correcthorsebatterystaple committed:
Add regression tests

lucasccordeiro commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -272,8 +274,16 @@ bool solidity_convertert::get_var_decl(
+    if(
+      SolidityGrammar::get_expression_t(init_value) ==
+      SolidityGrammar::ExpressionT::Literal)
+    {
```
There is no need for this bracket since we have just one instruction under the if-statement.

lucasccordeiro commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -12,6 +12,8 @@
 #include <util/message/format.h>
 #include <fstream>
 
+int DEBUG_COUNTER = 0;
```
Do we really need it?

mikhailramalho replied:
This needs to be removed.

correcthorsebatterystaple replied:
Has been overlook, will remove

lucasccordeiro commented on regression/esbmc-solidity/uint_overflow/test.desc:
```diff
@@ -0,0 +1,4 @@
+CORE
+contract.solast
+--function func --contract contract.sol
+^VERIFICATION SUCCESSFUL$
```
Can you also add a version of this regression test that produces `VERIFICATION FAILED`?

lucasccordeiro commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -711,6 +721,14 @@ bool solidity_convertert::get_statement(
+  return get_expr(expr, nullptr, new_expr);
+}
+
+bool solidity_convertert::get_expr(
```
Add some doxygen documentation?

lucasccordeiro commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -1400,6 +1457,39 @@ bool solidity_convertert::get_array_to_pointer_type(
   return false;
 }
 
+bool get_elementary_type_name_uint(
```
Add some documentation?

kunjsong01 replied:
make it a private function in class solidity_convertert?

correcthorsebatterystaple committed:
Add Docstrings for get_expr

kunjsong01 commented on src/solidity-frontend/solidity_grammar.h:
```diff
@@ -50,6 +50,37 @@ enum ElementaryTypeNameT
 {
   // rule unsigned-integer-type
   UINT8 = 0,
+  UINT16 = 1,
```
`= 1,` is not needed here. Same for the other items in this enum.

kunjsong01 commented on src/solidity-frontend/solidity_grammar.cpp:
```diff
@@ -65,21 +66,28 @@ TypeNameT get_type_name_t(const nlohmann::json &type_name)
   {
     // for AST node that contains ["typeName"]["typeDescriptions"]
-    if(type_name["typeString"] == "uint8" || type_name["typeString"] == "bool")
+    std::set<std::string> uintTypeStringSet = {
```
make it `static` in namespace SolidityGrammar?

kunjsong01 commented on src/solidity-frontend/solidity_grammar.cpp:
```diff
@@ -155,22 +163,34 @@ const char *type_name_to_str(TypeNameT type)
-    return UINT8;
-  }
-  else if(type_name["typeString"] == "bool")
+  std::map<std::string, ElementaryTypeNameT> uintMap = {
```
make it static in namespace SolidityGrammar?

kunjsong01 commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -1400,6 +1457,39 @@ bool solidity_convertert::get_array_to_pointer_type(
+  SolidityGrammar::ElementaryTypeNameT &type,
+  typet &out)
+{
+  std::map<SolidityGrammar::ElementaryTypeNameT, unsigned int> uintSizeMap = {
```
make it `static` in namespace SolidityGrammar?

kunjsong01 commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -1400,6 +1457,39 @@ bool solidity_convertert::get_array_to_pointer_type(
+  };
+
+  if(!uintSizeMap.count(type))
+    return true;
```
To make future debugging easier, would it be possible to report error and abort? e.g. `msg("type not found")`?

kunjsong01 commented on src/util/c_types.h:
```diff
@@ -18,6 +18,7 @@ type2tc index_type2();
 typet int_type();
 type2tc int_type2();
 typet uint_type();
+typet uint_type(unsigned int);
```
wrap it in "#ifdef ENABLE_SOLIDITY_FRONTEND" ?

mikhailramalho replied:
This is not needed. Check how we build _ExtInt in the C frontend.

mikhailramalho commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -900,18 +953,22 @@ bool solidity_convertert::get_binary_operator_expr(
       return true;
 
-    if(get_expr(expr["rightHandSide"], rhs))
+    if(get_expr(expr["rightHandSide"], literalType, rhs))
```
why are you passing the LHS type with the RHS expression? Shouldn't it be the expr's type?

correcthorsebatterystaple replied:
The way the AST is structured is that if the expression is `uint8 i = 22`, this is passed in with the type on the RHS to be just a generic type of int_const without any indication to which variant of the int/uint it is. We make the assumption that the literal on the right will share the type of the rhs. which is why we pass in that literal type.

mikhailramalho replied:
Right, but shouldn't you get the type of expression, instead of the RHS? Something like `expr["typeDescription"]`?

mikhailramalho commented on src/util/c_types.cpp:
```diff
@@ -111,6 +111,11 @@ typet uint_type()
   return unsignedbv_typet(config.ansi_c.int_width);
 }
 
+typet uint_type(unsigned int size)
```
This is not needed. Check how we build _ExtInt in the C frontend.

correcthorsebatterystaple replied:
Doesn't _ExtInt also do the same thing just directly with the `unsignedbv_typet`?

rafaelsamenezes replied:
The issue here is that `c_types` regards... C base types. Here you are basically creating an alias. Which is not needed.

correcthorsebatterystaple replied:
This makes sense, I have a redundant alias and as this is for solidity there is no reason for dependence on c_types. Will amend to use the unsignedbv_typet.

mikhailramalho commented on src/solidity-frontend/solidity_grammar.cpp:
```diff
@@ -155,22 +163,34 @@ const char *type_name_to_str(TypeNameT type)
     // For Literal, their typeString is like "int_const 100".
     // TODO: Fix me! For simplicity, we assume everything is unsigned int.
-    return UINT8;
+    return UINT256;
```
why 256? The comment seems to say otherwise.

correcthorsebatterystaple replied:
This will actually be removed as this PR fixes the defaulting

mikhailramalho commented on src/solidity-frontend/solidity_grammar.cpp:
```diff
@@ -155,22 +163,34 @@ const char *type_name_to_str(TypeNameT type)
+  {
+    return uintMap[typeString];
+  }
+  else if(typeString == "bool")
```
there is no need for the elses here since every if returns.

mikhailramalho reviewed (changes requested):
TBH, I don't understand this PR:
1. `convert_integer_literal` takes an `const nlohmann::json &integer_literal` but somehow we are passing a `literal_type`?
2. Why do we need the type after all? Since you're passing the expr around, you should be able to retrieve it whenever you want.

correcthorsebatterystaple commented:
@mikhailramalho 
> TBH, I don't understand this PR:
> 
> 1. `convert_integer_literal` takes an `const nlohmann::json &integer_literal` but somehow we are passing a `literal_type`?
> 2. Why do we need the type after all? Since you're passing the expr around, you should be able to retrieve it whenever you want.

1. This is a result of the existing structure, `convert_integer_literal` seems to also handle the creating of of the variables of uint types. The way the AST is built is that in an initialization with a literal rvalue `uint8 i = 10` we get type definitions for the left side but for the right side we get a generic integer type without any indication for what subtype of the integer it is (unsigned or signed, bitsize) which then requires us to pass in the type from the left hand side which is what the literal_type stores here
2. This is true that we are passing the expression about but because literal expressions do not contain any specific type descriptions we need to pass them around from the lhs of the expression.

correcthorsebatterystaple committed:
PR requested changes

- Add helper method get_elementary_type_name_uint to private member
- Add failure assertion when size UINT type not found
- Add ENUM_TO_STR for UINT types

correcthorsebatterystaple committed:
Add Logical and and verification failed test for uints

correcthorsebatterystaple committed:
Move uint maps to static in namespace

correcthorsebatterystaple committed:
Merge remote-tracking branch 'origin/master' into feature/add-solidity-uint-types

lucasccordeiro reviewed (approved):
LGTM.

lucasccordeiro commented:
@fbrausse and @rafaelsamenezes: could you please review this PR?

correcthorsebatterystaple committed:
Introduce INT_LITERAL type instead of using UINT256

lucasccordeiro commented:
@correcthorsebatterystaple: could you please address @rafaelsamenezes' comments?

correcthorsebatterystaple committed:
Remove type alias function from c_types

fbrausse commented on src/solidity-frontend/solidity_grammar.h:
```diff
@@ -4,6 +4,7 @@
 #include <map>
 #include <string>
 #include <nlohmann/json.hpp>
+#include <set>
```
This seems unused in this header.

fbrausse commented on src/solidity-frontend/solidity_grammar.h:
```diff
@@ -67,6 +101,24 @@ enum ElementaryTypeNameT
 };
 ElementaryTypeNameT get_elementary_type_name_t(const nlohmann::json &type_name);
 const char *elementary_type_name_to_str(ElementaryTypeNameT type);
+static std::map<ElementaryTypeNameT, unsigned int> uintSizeMap = {
```
While I understand exporting a way to map from the enum to the type might be necessary, making this static declaration in a header file is not a good approach. I have two issues with this:
1. not every mapping needs to be a `std::map`; as the enum is layed out, this could just as well be an array. Or a switch.
2. Having this `static` in the header means that every translation unit that #include-s it also gets a local `uintSizeMap` symbol. That is wasteful.

So, the suggestion is to move this symbol declaration to the place where it is needed and if exporting a way to obtain the size for one of these enum values is required, to add a function for that.

Also: if you keep the std::map, can it be `const`?

lucasccordeiro commented:
@correcthorsebatterystaple: could you please rebase this PR with the master branch?

correcthorsebatterystaple committed:
Merge branch 'master' into feature/add-solidity-uint-types

fbrausse commented on src/solidity-frontend/solidity_grammar.cpp:
```diff
@@ -65,21 +85,19 @@ TypeNameT get_type_name_t(const nlohmann::json &type_name)
-    if(type_name["typeString"] == "uint8" || type_name["typeString"] == "bool")
+    std::string typeString = type_name["typeString"].get<std::string>();
+
+    if(uintTypeStringSet.count(typeString) || typeString == "bool")
```
It seems that this is the only place `uintTypeStringSet` is used. Basically for checking whether the `typeString` matches one of these constants. Wouldn't it be possible to use `uintStringToTypeMap` for that purpose, since its keys are the same as in the set? Then the set could be removed.

correcthorsebatterystaple replied:
This is true, I will substitute the set checks with the map checks

fbrausse commented on src/solidity-frontend/solidity_grammar.cpp:
```diff
@@ -17,6 +17,26 @@ Author: Kunjian Song, kunjian.song@postgrad.manchester.ac.uk
+  "uint176", "uint184", "uint192", "uint200", "uint208", "uint216", "uint224",
+  "uint232", "uint240", "uint248", "uint256",
+};
+std::map<std::string, ElementaryTypeNameT> uintStringToTypeMap = {
```
Same as the comment in the header: can't this be `const`? Also, why is this symbol `uintStringToTypeMap` exported in the `SolidityGrammar` namespace? Couldn't it be local to this translation unit by defining it outside the namespace as `static`?

fbrausse reviewed (commented):
Some remarks about the declaration of symbols. As a general question: why did you choose `std::map` over `std::unordered_map`?

correcthorsebatterystaple committed:
Make maps unordered and replace map export with method

fbrausse commented on src/solidity-frontend/solidity_grammar.cpp:
```diff
@@ -198,6 +260,14 @@ const char *elementary_type_name_to_str(ElementaryTypeNameT type)
+  if(!uint_size_map.count(type))
+    assert(!"Size map missing for type");
+
+  return uint_size_map.at(type);
```
Since this throws an exception when `type` is not in the map, you don't need the `.count()` check above.

mikhailramalho commented on src/solidity-frontend/solidity_grammar.cpp:
```diff
@@ -252,6 +260,14 @@ const char *elementary_type_name_to_str(ElementaryTypeNameT type)
+unsigned int uint_type_name_to_size(ElementaryTypeNameT type)
+{
+  if(!uint_size_map.count(type))
+    assert(!"Size map missing for type");
```
when building with -DNDEBUG flag, this assertion will be removed, and the if guard will be applied to the return statement. Please use msg.error + abort here.

fbrausse replied:
The `.at()` will throw in that case, so I don't think the msg.error + abort() is needed here.

correcthorsebatterystaple committed:
Remove check for uint size

correcthorsebatterystaple committed:
Remove redundant else ifs

fbrausse commented on src/solidity-frontend/solidity_grammar.cpp:
```diff
@@ -65,21 +91,19 @@ TypeNameT get_type_name_t(const nlohmann::json &type_name)
-    else if(
-      type_name["typeString"].get<std::string>().find("int_const") !=
-      std::string::npos)
+    else if(typeString.find("int_const") != std::string::npos)
```
Sorry, I'm not familiar with the Solidity language nor the translated JSON very much. Is it possible to define own type names there? Could I for instance define a `float` (if it exists) to be called `my_int_const`? If yes, then this condition using `.find()` will (probably wrongly) recognize my "typedef'ed" `float` as an `int_const`. Is that a problem that can occur with Solidity?

correcthorsebatterystaple replied:
I believe solidity doesn't support type aliases, so these type strings are from a predefined set of keywords and cannot be user defined.

fbrausse commented on src/solidity-frontend/solidity_grammar.cpp:
```diff
@@ -65,21 +91,19 @@ TypeNameT get_type_name_t(const nlohmann::json &type_name)
-    else if(
-      type_name["typeString"].get<std::string>().find("function") !=
-      std::string::npos)
+    else if(typeString.find("function") != std::string::npos)
```
Same as above, is there a way to define a type called `my_function` in Solidity?

fbrausse commented on src/solidity-frontend/solidity_grammar.cpp:
```diff
@@ -155,31 +179,35 @@ const char *type_name_to_str(TypeNameT type)
+
+  if(
+    typeString.substr(0, 4) == "uint" &&
+    uint_string_to_type_map.count(typeString))
```
Why is the first condition needed when `.count()` already compares for equality (actually equivalence)?

correcthorsebatterystaple replied:
This is true the first condition is redundant, I will amend this.

fbrausse commented on src/solidity-frontend/solidity_grammar.cpp:
```diff
@@ -155,31 +179,35 @@ const char *type_name_to_str(TypeNameT type)
-    return UINT8;
-  }
-  else
+  if(typeString.find("int_const") != std::string::npos)
```
Again `.find()`: `my_int_const`?

fbrausse reviewed (commented):
Some questions about the type identifiers.

correcthorsebatterystaple committed:
Remove substring check for uint type

correcthorsebatterystaple committed:
Merge branch 'master' into feature/add-solidity-uint-types

fbrausse reviewed (approved):
> I believe solidity doesn't support type aliases, so these type strings are from a predefined set of keywords and cannot be user defined.

I see, thank you. Using equality instead of a substring test in these case seems more robust (this set of keywords might still change in the future) and consistent since all the `map.count()` checks use equality. As this PR is not about these conditions, I guess I'm fine with that.

LGTM.

correcthorsebatterystaple commented:
> > I believe solidity doesn't support type aliases, so these type strings are from a predefined set of keywords and cannot be user defined.
> 
> I see, thank you. Using equality instead of a substring test in these case seems more robust (this set of keywords might still change in the future) and consistent since all the `map.count()` checks use equality. As this PR is not about these conditions, I guess I'm fine with that.
> 
> LGTM.

To elaborate the type definitions on the AST show up as `int_const 4` for the literal integer 4, which is why there is a substring test as we assure that the type is a literal by checking that there is an int_const in there. I do, however, agree that this check should be done more robustly by either checking the parent node and verifying by that node that it is a literal. However, with the current architecture the type definition is passed around instead of the parent node which would require a refactor on how the type definitions are evaluated.
